clients/pkg/promtail/targets/kafka/handler.go

```diff
@@ -0,0 +1,61 @@
+package kafka
+
+import (
+	"github.com/Shopify/sarama"
+	"github.com/go-kit/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+const (
+	defaultKafkaMessageKey  = "none"
+	labelKeyKafkaMessageKey = "__meta_kafka_message_key"
+)
+
+// KafkaTarget allows making message handling independent from the internals of the specific Target internals
+type KafkaTarget interface {
+	// RelabelConfig returns relabel configs
+	RelabelConfig() []*relabel.Config
+	// Labels returns labels
+	Labels() model.LabelSet
+	// Client returns EntryHandler
+	Client() api.EntryHandler
+	// UseIncomingTimestamp tells to use a timestamp from the incoming message
+	UseIncomingTimestamp() bool
+	// Session returns sarama consumer group session
+	Session() sarama.ConsumerGroupSession
+	// Logger returns logger
+	Logger() log.Logger
+}
+
+// messageHandler is a default message handler for a kafka target
+func messageHandler(message *sarama.ConsumerMessage, t KafkaTarget) {
```

Comment:
Good point 👍 

Initially, I wanted more control over message handling, which is why I decided to go this way.

Anyway, I tried to change the code and inject only the parser, it reduced the amount of the code, and I got rid of a few new interfaces and some copy-paste. Thanks, @kgeckhart for the idea 👍

Reply:
+1 on using an interface like `KafkaMessageParser` for injection purposes

Reply:
I added `MessageParser` to the Kafka module. So it will be`kafka.MessageParser` 😃 

```go
// MessageParser defines parsing for each incoming message
type MessageParser interface {
	Parse(message *sarama.ConsumerMessage) ([]string, error)
}
```